src/Chainweb/Version/Utils.hs

```diff
@@ -0,0 +1,434 @@
+, nextGraphChange
+, chainCountAt
+, diameterAt
+, degreeAt
+, chainIdsAt
+, randomChainIdAt
+, someChainIdAt
+, blockCountAt
+, globalBlockCountAt
+, globalBlockRateAt
+, isGraphChange
+
+-- * Chain Graph Properties By Cut Height
+, avgCutHeightAt
+, chainGraphsByCutHeight
+, chainGraphAtCutHeight
+, chainCountAtCutHeight
+, diameterAtCutHeight
+, degreeAtCutHeight
+, blockCountAtCutHeight
+, avgBlockHeightAtCutHeight
+, lastGraphChangeByCutHeight
+, nextGraphChangeByCutHeight
+
+-- * Expected Block and Cut Heights
+, expectedBlockHeightAfterSeconds
+, expectedCutHeightAfterSeconds
+, expectedBlockCountAfterSeconds
+, expectedGlobalBlockCountAfterSeconds
+) where
+
+import Chainweb.BlockHeight
+import Chainweb.Time
+
+import Data.Foldable
+import qualified Data.HashSet as HS
+import qualified Data.Map.Strict as M
+
+import GHC.Stack
+
+import Numeric.Natural
+
+import System.Random
+
+-- internal modules
+
+import Chainweb.Graph
+import Chainweb.Utils
+import Chainweb.Version
+
+-- -------------------------------------------------------------------------- --
+--  Utils
+
+limitHeight :: BlockHeight -> M.Map BlockHeight a -> M.Map BlockHeight a
+limitHeight h = M.takeWhileAntitone (h >=)
+{-# INLINE limitHeight #-}
+
+atHeight :: HasCallStack => BlockHeight -> M.Map BlockHeight a -> a
+atHeight h = snd . fromJuste . M.lookupLE h
+{-# INLINE atHeight #-}
+
+atCutHeight :: HasCallStack => CutHeight -> M.Map CutHeight a -> a
+atCutHeight h = snd . fromJuste . M.lookupLE h
+{-# INLINE atCutHeight #-}
+
+-- -------------------------------------------------------------------------- --
+-- Chain Graph Properties By Block Height
+
+-- | BlockHeight intervals for the chain graphs of a chainweb version
+--
+-- Post-condition:
+--
+-- @
+-- 0 == minimum $ M.keys $ chainGraphs v
+-- @
+--
+chainGraphs
+    :: HasChainwebVersion v
+    => v
+    -> M.Map BlockHeight ChainGraph
+chainGraphs = M.fromList . toList . chainwebGraphs . _chainwebVersion
+{-# INLINE chainGraphs #-}
+
+-- | BlockHeight intervals for the chain graphs of a chainweb version up to a
+-- given block height.
+--
+-- Post-condition:
+--
+-- @
+-- 0 == minimum $ M.keys $ chainGraphs v
+-- @
+--
+chainGraphsAt
+    :: HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> M.Map BlockHeight ChainGraph
+chainGraphsAt v h = limitHeight h (chainGraphs v)
+{-# INLINE chainGraphsAt #-}
+
+lastGraphChange
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> BlockHeight
+lastGraphChange v h = fst . fromJuste . M.lookupLE h $ chainGraphs v
+{-# INLINE lastGraphChange #-}
+
+nextGraphChange
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> BlockHeight
+nextGraphChange v h = fst . fromJuste . M.lookupGT h $ chainGraphs v
+{-# INLINE nextGraphChange #-}
+
+chainCountAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> Natural
+chainCountAt v h = order $ atHeight h $ chainGraphs v
```

Comment:
Suggest refactor `atHeight h $ chainGraphs v` to shared function and use in the 4 places here

Reply:
hmm, I tried it, but decided that it wasn't worth it: `chainGraphsAtHeight v h` is not that much different than `atHeight h $ chainGraphs v` but adds an level of indirection when understanding the code.